demos/terraform/main.tf

```diff
@@ -0,0 +1,212 @@
+      uplink_device_name   = var.uplink_device_name
+      ovn_uplink_interface = var.ovn_uplink_interface
+      microceph_channel    = var.microceph_channel
+      microovn_channel     = var.microovn_channel
+      microcloud_channel   = var.microcloud_channel
+      lxd_channel          = var.lxd_channel
+    })
+  }
+
+  limits = {
+    cpu    = var.cpu_per_instance
+    memory = "${var.memory_per_instance}GiB"
+  }
+
+  device {
+    name = var.lookup_interface
+    type = "nic"
+    properties = {
+      nictype        = "bridged"
+      parent         = var.lookup_bridge
+      "ipv4.address" = local.systems[count.index].ip
+    }
+  }
+
+  device {
+    name = var.uplink_device_name
+    type = "nic"
+    properties = {
+      nictype = "bridged"
+      parent  = lxd_network.microbr0.name
+    }
+  }
+
+  device {
+    name = "local"
+    type = "disk"
+    properties = {
+      pool   = var.storage_pool
+      source = lxd_volume.local_disk[count.index].name
+    }
+  }
+
+  dynamic "device" {
+    for_each = local.systems[count.index].has_ceph ? [1] : []
+    content {
+      name = "remote"
+      type = "disk"
+      properties = {
+        pool   = var.storage_pool
+        source = lxd_volume.ceph_disk[local.ceph_disk_mapping[count.index]].name
+      }
+    }
+  }
+
+  file {
+    content = templatefile("${path.module}/microcloud_preseed.yaml.tpl", {
+      subnet                = local.lookup_subnet
+      internal_subnet       = local.lookup_subnet
+      lookup_timeout        = var.lookup_timeout
+      session_passphrase    = random_password.session_passphrase.result
+      session_timeout       = var.session_timeout
+      systems               = local.systems
+      ovn_ipv4_gateway      = var.ovn_ipv4_gateway
+      ovn_ipv4_range        = var.ovn_ipv4_range
+      ovn_ipv6_gateway      = var.ovn_ipv6_gateway
+      ovn_dns_servers       = var.ovn_dns_servers
+      local_disk_device     = var.local_disk_device
+      ceph_disk_device      = var.ceph_disk_device
+      uplink_device_name    = var.uplink_device_name
+      initiator             = var.initiator
+      ovn_uplink_interface  = var.ovn_uplink_interface
+      current_node_has_ceph = local.systems[count.index].has_ceph
+    })
+    target_path = "/root/microcloud_preseed.yaml"
+    mode        = "0600"
+  }
+
+
+  # cloud-init status --wait fails with an "invalid argument" error (129) because the command is triggered by Terraform's on_start hook very early in the boot cycle.
+  # At this point, the cloud-init daemon or its environment is not yet fully initialized to correctly respond to status queries, even though cloud-init eventually completes successfully.
+  #
+  # Solution: Wait for systemctl first to ensure the system is fully booted before calling cloud-init status --wait.
+  # This prevents the early boot timing issue and allows cloud-init to respond properly.
+
+
+  execs = var.vm_names[count.index] == var.initiator ? {  
```

Comment:
I ran the file through `terraform fmt` and it found some extraneous whitespace at the end of this line.